doc/sphinx/proof-engine/ltac.rst

```diff
@@ -633,10 +741,41 @@ Coq provides a derived tactic to check that a tactic *fails*:
    .. exn:: Tactic failure: <tactic closure> succeeds.
       :undocumented:
 
+   .. note:: :tacn:`assert_fails` and :tacn:`assert_succeeds` work as described when
+      :token:`ltac_expr3` is a :token:`simple_tactic`.  In some more complex expressions,
+      they may report an error from within :token:`ltac_expr3` when they shouldn't.
+      This is due to the order in which parts of the :token:`ltac_expr3`
+      are evaluated and executed.  For example:
+
+      .. coqtop:: reset none
+
+         Goal True.
+
+      .. coqtop:: all fail
+
+         assert_fails match True with _ => fail end.
+
+      should not show any message.  The issue is that :tacn:`assert_fails` is an |Ltac|-defined
+      tactic.  That makes it a function that's processed in the evaluation phase, causing
```

Comment:
It's actually a tactic notation, but that probably doesn't matter to the presentation here.  (The important thing is that it's not a special primitive construct like `tryif` or `+`.)

Reply:
It's a tactic notation that refers to an |Ltac|-defined tactic.  Good enough as it is.

BTW, it's odd to see `assert_succeeds` defined in terms of `assert_fails`.  

```
Ltac assert_succeeds tac :=
  tryif (assert_fails tac) then gfail 0 tac "fails" else idtac.
```

Reply:
> BTW, it's odd to see `assert_succeeds` defined in terms of `assert_fails`.

Yes, this is because there's no way to roll back the effects of a tactic except via making it fail.  We could do something more complicated, but it's much easier to just negate `assert_fails`.